### PR TITLE
Optional characteristics for lightbulbs

### PR DESCRIPTION
**Pull request by ccutrer**

### CODE REVIEW

**Comment by timcharper** [comment]
@ccutrer Thanks a lot for this change! This seems to be quite in line with my proposal for #63. @beowulfe how do you feel about that proposal and this change? I'd really like optional characteristics to not require some trait for the reasons outlined in #63.

**Comment by timcharper** [comment]
@beowulfe I'd really like to get your opinion on this; I'm trying to work towards integrating my openhab2 homekit changes merged in to master; if you think this can make a 1.1.x release, then I'll base my changes on it :)

**Comment by timcharper** [comment]
@beowulfe gentle nudge here! :) I know you're super busy, if there's anything I can do to help get this stuff merged and a release out, let me know. I'm just eager to integrate my openhab contributions before they diverge too far from master.

**Comment by timcharper** [comment]
@ccutrer looks like this breaks backwards compat. Let's hold the merge until we release another 1.1.x version, and then we can bump the base version to 1.2.x and break all the things.

File: src/main/java/com/beowulfe/hap/accessories/DimmableLightbulb.java [hunk 1]
```diff
@@ -1,38 +1,17 @@
...
-   * Subscribes to changes in the brightness of the light.
-   *
-   * @param callback the function to call when the state changes.
-   */
-  void subscribeBrightness(HomekitCharacteristicChangeCallback callback);
-
-  /** Unsubscribes from changes in the brightness of the light. */
-  void unsubscribeBrightness();
+@Deprecated
+public interface DimmableLightbulb extends Lightbulb, Brightness {
```
Comment (timcharper) [inline src/main/java/com/beowulfe/hap/accessories/DimmableLightbulb.java]: I've been thinking about it... we should really just do away with all of these interfaces and have of the allowed (but optional) characteristics in the Lightbulb itself.

**Comment by ccutrer** [comment]
Totes. That was all an attempt to maintain backwards compat (which didn't work out). Breaking that will make it all a lot easier.

**Comment by ccutrer** [comment]
PR updated with just the lightbulb refactor, removing the extra interfaces. I'll separate fan into a new PR.

File: src/main/java/com/beowulfe/hap/accessories/ColorfulLightbulb.java [hunk 2]
```diff
@@ -1,65 +1,18 @@
...
-import com.beowulfe.hap.HomekitCharacteristicChangeCallback;
-import java.util.concurrent.CompletableFuture;
+import com.beowulfe.hap.accessories.characteristics.Color;
+import java.util.Optional;
 
 /**
- * Extends {@link Lightbulb} with color settings. This will usually be implemented along with {@link
- * DimmableLightbulb}, but not necessarily so.
+ * Extends {@link com.beowulfe.hap.accessories.Lightbulb} with color settings. This will usually be
+ * implemented along with {@link DimmableLightbulb}, but not necessarily so.
```
Comment (ccutrer) [inline src/main/java/com/beowulfe/hap/accessories/ColorfulLightbulb.java]: ooh, this java doc needs updated.

File: src/main/java/com/beowulfe/hap/accessories/characteristics/Color.java [hunk 3]
```diff
@@ -1,16 +1,15 @@
...
- * Extends {@link Lightbulb} with color settings. This will usually be implemented along with {@link
- * DimmableLightbulb}, but not necessarily so.
+ * Extends a Lightbulb with color settings. This will usually be implemented along with {@link
+ * Brightness}, but not necessarily so.
  *
  * @author Andy Lintner
  */
-public interface ColorfulLightbulb extends Lightbulb {
-
+public interface Color {
```
Comment (timcharper) [inline src/main/java/com/beowulfe/hap/accessories/characteristics/Color.java]: I like where this is going. After some thought, I wonder if this interface is probably more trouble than it's worth.

Perhaps we can just axe it and have the respective characteristics take lambda functions as appropriate. For read-only characteristics, it takes only a getter function and a subscription. For write-only, only a setter. Then the respective accessories can just return an Optional of the characteristic, with the lambda functions defined as needed?

WDYT?

Comment (ccutrer) [reply src/main/java/com/beowulfe/hap/accessories/characteristics/Color.java]: Yup, I like that.

**Pull request closed without merging**